### ISSUE TITLE
how to get the scores

### ISSUE THREAD

**Issue by rippleD030** [facebookresearch/chai#4]
Hello,

I would like to ask how the scores used for KMeans clustering in perform_offline_clustering.py are obtained through the main_kmeans_avg function. Are these scores the result of `applying scores = F.softmax(scores.float(), dim=-1).type_as(xq)` during each forward pass of the model's attention layer? If that’s the case, I find it challenging to filter out the zero values with the line 
```
first_head = scores_numpy[0, :] 
num_nonzero = first_head[first_head != 0].shape[0] 
reshaped_data = scores_numpy[:, :num_nonzero]
```
which makes me feel like I might be doing something wrong. Could you please advise on the proper approach? Thank you very much, and I look forward to your response.

**Comment by bilgeacun** [comment]
Hello, you can use the following to dump the scores:
```
scores_numpy = scores.cpu().numpy()
with open(
     f"checkpoint/layer_{self.layer_id}_scores.jsonl",
     "a",
) as fout:
     for example_id in range(scores.shape[0]):
         # for head_num in range(scores.shape[1]):
         for seq_actv in range(scores.shape[2]):
             out_value = (
                 scores[example_id, :, seq_actv, :].cpu().tolist()
             )
             json.dump(out_value, fout)
             fout.write("\n")
```

And use perform_offline_clustering.py with providing the dump path as:
`python perform_offline_clustering.py --path <dump_path> `
